lib/ansible/module_utils/api.py

```diff
@@ -91,26 +92,57 @@ def ratelimited(*args, **kwargs):
     return wrapper
 
 
-def retry(retries=None, retry_pause=1):
-    """Retry decorator"""
-    def wrapper(f):
+def _retry_limit_exceeded(retries):
+    def function_wrapper(function):
+        def raise_error(*args, **kwargs):
+            raise Exception("Retry limit exceeded: %d" % retries)
+        return raise_error
+    return function_wrapper
 
-        def retried(*args, **kwargs):
-            retry_count = 0
-            if retries is not None:
-                ret = None
-                while True:
-                    retry_count += 1
-                    if retry_count >= retries:
-                        raise Exception("Retry limit exceeded: %d" % retries)
-                    try:
-                        ret = f(*args, **kwargs)
-                    except Exception:
-                        pass
-                    if ret:
-                        break
-                    time.sleep(retry_pause)
-                return ret
-
-        return retried
-    return wrapper
+
+def retry(retries=None, retry_pause=1):
+    """Retry decorator.
+    This assumes the function has already been called, so if retries == 1, the maximum attempts
+    have been reached. If retries is None, the function is a no-op.
+    """
+    if retries is None:
+        return lambda function: lambda *args, **kwargs: None
+    if retries <= 1:
+        return _retry_limit_exceeded(retries)
+    try:
+        backoff_delays = [retry_pause for delay in range(0, retries)]
+        return retry_wrapper(backoff_iterator=backoff_delays, retry_condition=lambda x: True)
+    except Exception:
```

Comment:
This turns any error into a broad rate limit exception, since we control the emitted exception we can catch just it and any other exception wouldn't be silently shadowed.

Reply:
I'm not sure what you mean or what you'd like me to change.

Reply:
I'd like `except RetryError` I think.

Reply:
Okay, good idea :+1:

Reply:
I've *refactored* this function, but not intended to change behavior. Anyway, I'm going to just remove the refactor. No need to backport this for sure, and I don't think the refactor is worth much anyway :smile: